tests/MiddlewareDispatcherTest.php

```diff
@@ -0,0 +1,119 @@
+<?php
+
+
+namespace Yiisoft\Yii\Web\Tests;
+
+use PHPUnit\Framework\TestCase;
+use Psr\Container\ContainerInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use Yiisoft\Di\Container;
+use PHPUnit_Framework_MockObject_MockObject;
+use Yiisoft\Yii\Web\Emitter\SapiEmitter;
+use Yiisoft\Yii\Web\MiddlewareDispatcher;
+
+class MiddlewareDispatcherTest extends TestCase
+{
+    /**
+     * @var MiddlewareDispatcher
+     */
+    private $middlewareDispatcher;
+
+    /**
+     * @var Container|PHPUnit_Framework_MockObject_MockObject
+     */
+    private $containerMock;
+
+    /**
+     * @var RequestHandlerInterface|PHPUnit_Framework_MockObject_MockObject
+     */
+    private $fallbackHandlerMock;
+
+    /**
+     * @var MiddlewareInterface[]|PHPUnit_Framework_MockObject_MockObject[]
+     */
+    private $middlewareMocks;
+
+    public function setUp()
+    {
+        parent::setUp();
+        $this->containerMock = $this->createMock(ContainerInterface::class);
+        $this->fallbackHandlerMock = $this->createMock(RequestHandlerInterface::class);
+        $this->middlewareMocks = [
+            $this->createMock(MiddlewareInterface::class),
+            $this->createMock(MiddlewareInterface::class)
+        ];
+        $this->middlewareDispatcher = new MiddlewareDispatcher($this->middlewareMocks, $this->containerMock, $this->fallbackHandlerMock);
+    }
+
+    /**
+     * @test
+     */
+    public function addThrowsInvalidArgumentExceptionWhenMiddlewareIsNotOfCorrectType()
+    {
+        $this->expectException(\InvalidArgumentException::class);
+        $exampleInput = new SapiEmitter();
+
+        $this->middlewareDispatcher->add($exampleInput);
+    }
+
+    /**
+     * @test
+     * @doesNotPerformAssertions
+     */
+    public function addAddsCallableToMiddlewareArrayWithoutThrowingException()
+    {
+        $callable = function () {
+            echo 'example function for testing purposes';
+        };
+        $this->middlewareDispatcher->add($callable);
+    }
+
+    /**
+     * @test
+     * @doesNotPerformAssertions
+     */
+    public function addAddsMiddlewareInterfaceToMiddlewareArrayWithoutThrowingException()
+    {
+        $middleware = $this->createMock(MiddlewareInterface::class);
+        $this->middlewareDispatcher->add($middleware);
+    }
+
+    /**
+     * @test
+     */
+    public function handleCallsFallbackHandlerWhenPointerValueEqualsMiddlewareQueueSize()
```

Comment:
> WhenPointerValueEqualsMiddlewareQueueSize

That's internal detail that really means fallback handler is called if no middleware returned response. I'd use meaning in the test name instead of internal details.